engine/table/src/main/java/io/deephaven/engine/util/OuterJoinTools.java

```diff
@@ -0,0 +1,410 @@
+     * table1 if the columnsToAdd is length zero). The returned table will have one row for each matching set of keys
+     * between the first and second tables, plus one row for any first table key set that doesn't match the second table
+     * and one row for each key set from the second table that doesn't match the first table. Columns from either table
+     * for which there was no match in the other table will have null values. Note that this method will cause tick
+     * expansion with ticking tables.
+     * <p>
+     *
+     * @param table1 input table
+     * @param table2 input table
+     * @param columnsToMatch match criteria
+     * @param columnsToAdd columns to add
+     * @return a table that has one column for each of table1's columns, and one column corresponding to each of
+     *         table2's columns listed in the columns to add (or all the columns whose names don't overlap with the name
+     *         of a column from table1 if the columnsToAdd is length zero). The returned table will have one row for
+     *         each matching set of keys between the first and second tables, plus one row for any first table key set
+     *         that doesn't match the second table and one row for each key set from the second table that doesn't match
+     *         the first table. Columns from either table for which there was no match in the other table will have null
+     *         values.
+     */
+    @ScriptApi
+    public static Table fullOuterJoin(
+            @NotNull final Table table1,
+            @NotNull final Table table2,
+            @NotNull final MatchPair[] columnsToMatch,
+            @NotNull final MatchPair[] columnsToAdd) {
+        // perform the leftOuterJoin; it's missing right-side only rows
+        final Table leftTable = leftOuterJoin(table1, table2, columnsToMatch, columnsToAdd);
+
+        // find a sentinel column name to use to identify right-side only rows
+        int numAttempts = 0;
+        String sentinelColumnName;
+        final Map<String, ? extends ColumnSource<?>> resultColumns = leftTable.getColumnSourceMap();
+        do {
+            sentinelColumnName = "__sentinel_" + (numAttempts++) + "__";
+        } while (resultColumns.containsKey(sentinelColumnName));
+
+        // handle the simpler scenario where no conflicts may exist otherwise leftOuterJoin would have failed
+        if (columnsToAdd.length == 0) {
+            final MatchPair[] matchPairs = Arrays.stream(columnsToMatch)
+                    .map(mp -> new MatchPair(mp.rightColumn(), mp.leftColumn()))
+                    .toArray(MatchPair[]::new);
+
+            final Table leftWithSentinel = table1.coalesce().updateView(sentinelColumnName + " = true");
+
+            // we must use leftOuterJoin again as there may be multiple matches per group
+            final Table rightTable = CrossJoinHelper.leftOuterJoin(
```

Comment:
So this is an interesting take on it; because we're going to end up doing the full cross join and then ditching part of the answer.

We are eliminating one hash table for the naturalJoin (or whereIn), which should be a win; the question is whether the necessary cross join state is worse is open and may depend on the data.

The columns are also out of order; but I guess merge handles that nicely?

Reply:
The merge does handle the out of order columns by respecting the ordering in the first table.

I agree the cross-join output may be terrible (and in specific scenarios absolutely is). Is it worse than the extra memory usage? I don't know; it's hard to tell and probably is sometimes and probably isn't sometimes.

Best solution probably involves incrementally improving the cross join mode to be an enum of INNER_JOIN, LEFT_OUTER_JOIN, and FULL_OUTER_JOIN. This is likely the biggest performance win for effort.

I'll chat offline with you about what you/@rcaudy prefer as part of this PR.